File: java/Components/Strata.Domain.Core/src/strata/domain/core/domainevent/AbstractDomainEvent.java
//////////////////////////////////////////////////////////////////////////////
// AbstractDomainEvent.java
//////////////////////////////////////////////////////////////////////////////

package strata.domain.core.domainevent;

import java.time.Instant;
import java.util.UUID;

/*****************************************************************************
 * Base implementation of {@code IDomainEvent<S>} interface.
 *
 * @param <S> event source type
 */
public abstract
class AbstractDomainEvent<S>
    implements IDomainEvent<S>
{
    private final String  name;
    private final UUID    correlationId;
    private final Instant timestamp;
    private final S       source;

    /*************************************************************************
     * Creates a new instance of {@code AbstractDomainEvent<S>}.
     *
     * @param nm  event name
     * @param src event source
     */
    protected
    AbstractDomainEvent(String nm,S src)
    {
        this(nm,null,src);
    }

    /*************************************************************************
     * Creates a new instance of {@code AbstractDomainEvent<S>}.
     *
     * @param nm  event name
     * @param correlId correlation id
     * @param src event source
     */
    protected
    AbstractDomainEvent(String nm,UUID correlId,S src)
    {
        name = nm;
        correlationId = correlId;
        timestamp = Instant.now();
        source = src;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public String
    getName()
    {
        return name;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public UUID
    getCorrelationId()
    {
        return correlationId;
    }

    /*************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    public Instant
    getTimestamp()
    {
        return timestamp;
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    getSource()
    {
        return source;
    }
}

//////////////////////////////////////////////////////////////////////////////

File: csharp/Components/Strata.Domain.Core/Shared/AbstractDomainEvent.cs
//  ##########################################################################
//  # File Name: AbstractDomainEvent.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;

namespace Strata.Domain.Core.Shared
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public abstract
    class AbstractDomainEvent<S>:
        IDomainEvent<S>
    {
        public virtual string   Name { get; protected set; }
        public virtual DateTime Timestamp { get; protected set; }
        public virtual S        Source { get; protected set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractDomainEvent(string name,S source):
            this(name,DateTime.Now,source) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractDomainEvent(string name,DateTime timestamp,S source)
        {
            Name      = name;
            Timestamp = timestamp;
            Source    = source;
        }
    }
}

//  ##########################################################################
